ios/NotificationServiceExtension/NotificationService.swift

```diff
@@ -0,0 +1,239 @@
+    } else {
+      contentHandler(bestAttemptContent)
+    }
+  }
+  
+  /**
+   * Parses the notification content and modifies it to be a Communication Notification. More info here: https://developer.apple.com/documentation/usernotifications/implementing_communication_notifications
+   */
+  @available(iOSApplicationExtension 15.0, *)
+  func configureCommunicationNotification(notificationContent: UNMutableNotificationContent, contentHandler: @escaping (UNNotificationContent) -> Void) {
+    var notificationData: NotificationData
+    do {
+      notificationData = try parsePayload(notificationContent: notificationContent)
+    } catch ExpError.runtimeError(let errorMessage) {
+      os_log("[NotificationService] configureCommunicationNotification() - couldn't parse the payload '%@'", log: log, type: .error, errorMessage)
+      contentHandler(notificationContent)
+      return
+    } catch {
+      os_log("[NotificationService] configureCommunicationNotification() - unexpected error while parsing payload", log: log, type: .error)
+      contentHandler(notificationContent)
+      return
+    }
+    
+    // Create an intent for the incoming message
+    let intent: INSendMessageIntent = createMessageIntent(notificationData: notificationData)
+
+    // Use the intent to initialize the interaction.
+    let interaction = INInteraction(intent: intent, response: nil)
+
+
+    // Interaction direction is incoming because the user is
+    // receiving this message.
+    interaction.direction = .incoming
+
+
+    // Donate the interaction before updating notification content.
+    interaction.donate { error in
+        if error != nil {
+            os_log("[NotificationService] configureCommunicationNotification() - failed to donate the message intent", log: self.log, type: .error)
+            contentHandler(notificationContent)
+            return
+        }
+        
+        // After donation, update the notification content.
+        do {
+          // Update notification content before displaying the
+          // communication notification.
+          let updatedContent = try notificationContent.updating(from: intent)
+          
+          // Call the content handler with the updated content
+          // to display the communication notification.
+          contentHandler(updatedContent)
+        } catch {
+          os_log("[NotificationService] configureCommunicationNotification() - failed to update the notification with send message intent", log: self.log, type: .error)
+          contentHandler(notificationContent)
+        }
+    }
+  }
+  
+  func parsePayload(notificationContent: UNMutableNotificationContent) throws -> NotificationData  {
+    guard let payload = notificationContent.userInfo["payload"] as? NSDictionary else {
+      throw ExpError.runtimeError("payload missing")
+    }
+    
+    guard let reportID = payload["reportID"] as? Int64 else {
+      throw ExpError.runtimeError("payload.reportID missing")
```

Comment:
nit: Apart from logging, do we have any other purpose of these errors? I can see we've got a catch based on ExpError. The reason I ask is because

Another way I've generally used is defining appropriate error types. An example:

```
enum ExpError: Error {
    case missingPayload
    case missingParam(String)
}

guard let reportID = payload["reportID"] as? Int64 else {
    throw ExpError.missingValue("reportID")
}

```

Reply:
Yeah these ExpErrors are just for adding a message to the error and logging it. If we start writing more native Swift code, we can start thinking about errors and logging a bit more carefully. For now just a basic message should be good.